### PR TITLE
Feat/early stopping + burn train refactor

### LINKED ISSUE #841: Early stopping
Add the option to keep the best weighs ranked by validation loss.
Stop the training when the X following epochs don't improve the metric.

## Optional

Being able to set the metric to be used for ranking.

### PR DESCRIPTION
**Pull request by nathanielsimard**
Another big refactor, removing the messy event collector and introduce an event store and an event processor.

Fix #841

### CODE REVIEW

File: burn-train/src/learner/early_stopping.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
+use crate::metric::{
+    store::{Aggregate, Direction, EventStoreClient, Split},
+    Metric,
+};
+
+/// The condition that [early stopping strategies](EarlyStoppingStrategy) should follow.
+pub enum StoppingCondition {
+    /// When no improvement has happened since the given number of epochs.
+    /// In other words, when no best epoch has been found.
```
Comment (louisfd) [inline burn-train/src/learner/early_stopping.rs:9]: I feel like this second sentence is just confusing. The condition is already clear from the first sentence, and there's always a "best" epoch even if it stagnates

File: burn-train/src/learner/early_stopping.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    NoImprovementSince {
+        /// The number of epochs allowed to get worsen before it get better.
```
Comment (louisfd) [inline burn-train/src/learner/early_stopping.rs:11]: The number of epochs allowed to worsen before it gets better.

File: burn-train/src/learner/early_stopping.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            match store.find_metric(&self.metric_name, epoch, self.aggregate, self.split) {
+                Some(value) => value,
+                None => {
+                    log::warn!("Can't find metric for early stopping.");
+                    return false;
+                }
+            };
+
+        let is_best = match self.direction {
+            Direction::Lowest => current_value <= self.best_value,
```
Comment (louisfd) [inline burn-train/src/learner/early_stopping.rs:46]: Shouldn't those be _strictly_ less than / greater than?
Because now even if there is no improvement and stays the same, it will be considered as improvement

File: burn-train/src/learner/early_stopping.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        // Two points for the last epoch. Mean 1.5
+        epoch += 1;
+        process_train(&mut processor, 1.0, epoch);
+        process_train(&mut processor, 2.0, epoch);
+        end_epoch(&mut processor, epoch);
+
+        assert!(
+            early_stopping.should_stop(epoch, &store),
+            "Should stop since two following epochs are worse then the best."
+        );
```
Comment (louisfd) [inline burn-train/src/learner/early_stopping.rs:183]: Add a test for the equality case

**Review by louisfd — CHANGES_REQUESTED** [review]
This refactor is really becoming legendary! I feel we gain in both cleanness and flexibility. 
I have some comments concerning the StoppingCriterion part, the rest LGTM
